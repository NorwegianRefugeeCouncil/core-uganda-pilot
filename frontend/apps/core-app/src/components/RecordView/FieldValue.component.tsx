import * as React from 'react';
import { Stack, Text, Link, useBreakpointValue } from 'native-base';
import { FieldKind } from 'core-api-client';

import { NormalisedBasicField } from './RecordView.types';

type Props = {
  item: NormalisedBasicField;
};

const useResponsiveStyles = () => {
  const direction = useBreakpointValue({
    sm: 'column',
    md: 'row',
  });

  const alignItems = useBreakpointValue({
    sm: 'flex-start',
    md: 'flex-start',
  });

  const justifyContent = useBreakpointValue({
    sm: 'center',
    md: 'flex-start',
  });

  const space = useBreakpointValue({
    sm: '1',
    md: '2',
  });

  const labelWidth = useBreakpointValue({
    sm: '100%',
    md: '40%',
  });

  const valueWidth = useBreakpointValue({
    sm: '100%',
    md: '60%',
  });

  return {
    direction,
    alignItems,
    justifyContent,
    space,
    labelWidth,
    valueWidth,
  };
};

export const FieldValueComponent: React.FC<Props> = ({ item }) => {
  const styles = useResponsiveStyles();

  return (
    <Stack
      direction={styles.direction}
      alignItems={styles.alignItems}
      justifyContent={styles.justifyContent}
      space={styles.space}
      width="100%"
    >
      <Text variant="body" color="neutral.300" width={styles.labelWidth}>
        {item.label}
      </Text>
      {item.fieldType === FieldKind.Reference ? (
        <Link href={`/record/${item.value}`} width={styles.valueWidth}>
          {item.formattedValue}
        </Link>
      ) : (
        <Text variant="label" width="60%">
          {item.formattedValue}
        </Text>
      )}
    </Stack>
  );
};
