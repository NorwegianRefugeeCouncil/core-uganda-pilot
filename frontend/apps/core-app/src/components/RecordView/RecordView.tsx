import * as React from 'react';
import { Record } from 'core-api-client';

type Props = {
  record: Record;
};

export const RecordView: React.FC<Props> = ({ record }) => {
  return <>{record.id}</>;
};
