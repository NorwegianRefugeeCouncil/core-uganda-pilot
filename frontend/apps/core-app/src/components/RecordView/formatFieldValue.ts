import { FieldDefinition, FieldKind, getFieldKind } from 'core-api-client';

// This method takes a field value and returns a formatted string
// It handles extra cases, like a multi-select field value that is not an array, for safety
export const formatFieldValue = (
  value: string | string[] | null,
  field: FieldDefinition,
): string => {
  const isNull = value === null;
  const isArray = Array.isArray(value);

  switch (getFieldKind(field.fieldType)) {
    case FieldKind.Checkbox:
      return isArray || !value || value === 'false' ? 'No' : 'Yes';
    case FieldKind.Date:
    case FieldKind.Month:
    case FieldKind.Week:
      return isArray || isNull ? '-' : value;
    case FieldKind.Reference:
      // TODO: Lookup reference form?
      return isArray || isNull ? '-' : value;
    case FieldKind.SingleSelect:
      return !isArray
        ? field.fieldType.singleSelect?.options.find((o) => o.id === value)
            ?.name ?? '-'
        : '-';
    case FieldKind.MultiSelect:
      return isArray
        ? value
            .map(
              (v) =>
                field.fieldType.multiSelect?.options.find((o) => o.id === v)
                  ?.name ?? '-',
            )
            .join(', ')
        : field.fieldType.multiSelect?.options.find((o) => o.id === value)
            ?.name ?? '-';
    case FieldKind.SubForm:
      return 'Error: Subform fields should not reach this point';
    default: {
      if (isNull) return '-';
      if (isArray) return value.join(', ');
      return value.toString();
    }
  }
};
