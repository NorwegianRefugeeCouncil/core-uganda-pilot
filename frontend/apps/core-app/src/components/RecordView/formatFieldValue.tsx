import { FieldDefinition, FieldKind, getFieldKind } from 'core-api-client';

export const formatFieldValue = (
  value: string | string[] | null,
  field: FieldDefinition,
): string => {
  if (value === null) return '-';

  const isArray = Array.isArray(value);

  switch (getFieldKind(field.fieldType)) {
    case FieldKind.Checkbox:
      return value === 'true' ? 'Yes' : 'No';
    case FieldKind.Date:
    case FieldKind.Month:
    case FieldKind.Week:
      return !isArray ? value : '-';
    case FieldKind.Reference:
      // TODO: Lookup reference form?
      return !isArray ? value : '-';
    case FieldKind.SingleSelect:
      return !isArray
        ? field.fieldType.singleSelect?.options.find((o) => o.id === value)
            ?.name ?? '-'
        : '-';
    case FieldKind.MultiSelect:
      return isArray
        ? value
            .map(
              (v) =>
                field.fieldType.multiSelect?.options.find((o) => o.id === v)
                  ?.name ?? '-',
            )
            .join(', ')
        : '-';
    case FieldKind.SubForm:
      return 'Error: Subform fields should not reach this point';
    default:
      return isArray ? value.join(', ') : value.toString();
  }
};
