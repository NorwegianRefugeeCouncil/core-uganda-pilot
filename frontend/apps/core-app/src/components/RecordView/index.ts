export { RecordViewComponent as RecordView } from './RecordView.component';
