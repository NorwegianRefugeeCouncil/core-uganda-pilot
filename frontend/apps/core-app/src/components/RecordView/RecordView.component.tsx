import * as React from 'react';
import { VStack } from 'native-base';
import { FieldKind } from 'core-api-client';

import { SubformFieldValueComponent } from './SubformFieldValue.component';
import { NormalisedFieldValue } from './RecordView.types';
import { FieldValueComponent } from './FieldValue.component';

type Props = {
  fieldValues: NormalisedFieldValue[];
};

export const RecordViewComponent: React.FC<Props> = ({ fieldValues }) => {
  return (
    <VStack space={4}>
      {fieldValues.map((f, i) => {
        if (f.fieldType === FieldKind.SubForm) {
          return (
            <SubformFieldValueComponent
              key={i}
              header={f.header}
              labels={f.labels}
              items={f.values}
            />
          );
        }

        return <FieldValueComponent key={i} item={f} />;
      })}
    </VStack>
  );
};
