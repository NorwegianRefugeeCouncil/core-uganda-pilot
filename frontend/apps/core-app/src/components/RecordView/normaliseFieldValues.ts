import {
  FieldKind,
  FormDefinition,
  getFieldKind,
  Record,
} from 'core-api-client';

import { formatFieldValue } from './formatFieldValue';
import {
  NormalisedBasicField,
  NormalisedFieldValue,
  NormalisedSubFormField,
  NormalisedSubFormFieldValue,
} from './RecordView.types';

export const normaliseFieldValues = (
  form: FormDefinition,
  record: Record,
): NormalisedFieldValue[] =>
  record.values.reduce<NormalisedFieldValue[]>((acc, cur) => {
    const field = form.fields.find((f) => f.id === cur.fieldId);
    if (!field) return acc;
    const fieldType = getFieldKind(field.fieldType);

    if (fieldType === FieldKind.SubForm) {
      if (!Array.isArray(cur.value)) return acc;
      const values: NormalisedSubFormFieldValue[][] = cur.value.map(
        (subRecord) => {
          if (!Array.isArray(subRecord)) return [];
          return subRecord.map((subFieldValue) => {
            const subField = field.fieldType.subForm?.fields.find(
              (f) => f.id === subFieldValue.fieldId,
            );
            if (!subField) throw new Error('subField not found');
            const subFieldType = getFieldKind(subField.fieldType);
            if (subFieldType === FieldKind.SubForm)
              throw new Error('subField is a subform');
            return {
              value: subFieldValue.value as string | string[] | null,
              formattedValue: formatFieldValue(
                subFieldValue.value as string | string[] | null,
                subField,
              ),
              fieldType: subFieldType,
            };
          });
        },
      );

      const labels =
        field.fieldType.subForm?.fields.map((subField) => subField.name) ?? [];

      const item: NormalisedSubFormField = {
        fieldType,
        header: field.name,
        labels,
        values,
      };

      return [...acc, item];
    }

    const item: NormalisedBasicField = {
      label: field.name,
      value: cur.value as string | string[] | null,
      fieldType,
      formattedValue: formatFieldValue(
        cur.value as string | string[] | null,
        field,
      ),
    };

    return [...acc, item];
  }, []);
