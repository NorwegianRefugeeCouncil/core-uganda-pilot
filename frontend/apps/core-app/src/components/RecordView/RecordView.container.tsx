import * as React from 'react';
import { FormDefinition, Record } from 'core-api-client';

import { RecordViewComponent } from './RecordView.component';
import { normaliseFieldValues } from './normaliseFieldValues';

type Props = {
  form: FormDefinition;
  record: Record;
  hideKeyFields?: boolean;
};

export const RecordViewContainer: React.FC<Props> = ({
  form,
  record,
  hideKeyFields,
}) => {
  const fieldValues = normaliseFieldValues(form, record);

  return (
    <RecordViewComponent
      fieldValues={fieldValues}
      hideKeyFields={hideKeyFields}
    />
  );
};
