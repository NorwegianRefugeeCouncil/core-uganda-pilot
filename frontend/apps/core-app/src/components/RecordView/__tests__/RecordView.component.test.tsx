import { FieldKind } from 'core-api-client';

import { render } from '../../../testUtils/render';
import {
  NormalisedBasicField,
  NormalisedFieldValue,
  NormalisedSubFormFieldValue,
} from '../RecordView.types';
import { RecordViewComponent } from '../RecordView.component';

jest.mock('../FieldValue.component', () => {
  const { View, Text } = jest.requireActual('react-native');
  return {
    FieldValueComponent: ({ item }: { item: NormalisedBasicField }) => (
      <View>
        <Text>{JSON.stringify(item)}</Text>
      </View>
    ),
  };
});

jest.mock('../SubformFieldValue.component', () => {
  const { View, Text } = jest.requireActual('react-native');
  return {
    SubformFieldValueComponent: ({
      header,
      labels,
      items,
    }: {
      header: string;
      labels: string[];
      items: NormalisedSubFormFieldValue[][];
    }) => (
      <View>
        <Text>{header}</Text>
        <Text>{JSON.stringify(labels)}</Text>
        <Text>{JSON.stringify(items)}</Text>
      </View>
    ),
  };
});

it('should match the snapshot', () => {
  const fieldValues: NormalisedFieldValue[] = [
    {
      label: 'label-1',
      fieldType: FieldKind.Text,
      value: 'test-1',
      formattedValue: 'format-test-1',
    },
    {
      fieldType: FieldKind.SubForm,
      header: 'header-1',
      labels: ['label-1', 'label-2'],
      values: [
        [
          {
            fieldType: FieldKind.Text,
            value: 'test-2',
            formattedValue: 'format-test-2',
          },
        ],
      ],
    },
  ];

  const { toJSON } = render(<RecordViewComponent fieldValues={fieldValues} />);
  expect(toJSON()).toMatchSnapshot();
});
