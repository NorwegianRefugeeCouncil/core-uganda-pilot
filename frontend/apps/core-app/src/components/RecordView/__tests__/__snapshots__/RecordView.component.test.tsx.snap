// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    dataSet={Object {}}
    style={Object {}}
  >
    <View>
      <Text>
        {"label":"label-1","fieldType":"text","value":"test-1","formattedValue":"format-test-1"}
      </Text>
    </View>
    <View
      dataSet={Object {}}
      style={
        Object {
          "height": 16,
        }
      }
    />
    <View>
      <Text>
        header-1
      </Text>
      <Text>
        ["label-1","label-2"]
      </Text>
      <Text>
        [[{"fieldType":"text","value":"test-2","formattedValue":"format-test-2"}]]
      </Text>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
